{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mock Data - fastq\n",
    "By Joely Nelson\n",
    "\n",
    "Contact: joelyn@cs.washington.edu\n",
    "\n",
    "The purpose of this file is to generate mock fastq files based on the mock bin data\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The illumina sequencer does not produce a nice xlsx file with the sequences already nicely bins. Instead it will give us fastq files. This file is intended to use the binned files to creat mock fastq files.\n",
    "\n",
    "#### Pipeline Visalization\n",
    "![](.\\\\documentation\\\\mock_fastq_pipeline.png)\n",
    "\n",
    "#### Parameters\n",
    "This file takes in the following parameters:\n",
    "\n",
    "* ``seq_name``: the file name to be read. It should have a column \"ID\" which is unique for every guide variant, and a \"Sequence\" column which contain's that guide's genetic sequence.\n",
    "* ``barcode_fname``: the txt file that contains all the barcodes, where each barcode corresponds to each bin. Each barcode is on it's own line.\n",
    "* ``bin_fname``: the file name that has the generated mock bins\n",
    "\n",
    "#### Output\n",
    "\n",
    "First, for each guide variant in the the associated read1 and read 2 are generated.\n",
    "\n",
    "read 1 is (39 bases; begins with: TAGG (TATAATACTAGT) N20 GTT) and read2 is (36 bases: reverse complement of N20, reverse complement of (sequence) CCTA).\n",
    "\n",
    "Then a unique barcode is assigned to each of the 10 bins.\n",
    "\n",
    "Two fastq files are generated:\n",
    "\n",
    "```read1_barcode_as_index.fastq```\n",
    "\n",
    "and\n",
    "\n",
    "```read2_barcode_as_index.fastq```\n",
    "\n",
    "* Used the documentation provided here: https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/FileFormat_FASTQ-files_swBS.htm to generate the files\n",
    "* Each file of the two files contains 25980000 records. They have placeholders (like the phred score and run number and all that) for all elements except for:\n",
    "    * read: is 1 or 2 depending on if it is read1 or read2\n",
    "    * index: the barcode index associated with bin\n",
    "    * sequence: the sequence of the guide variant based\n",
    "    \n",
    "#### Example    \n",
    "Take poxBp_T7_TGT. It has a sequence of:\n",
    "\n",
    "```\n",
    "CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCTAGCTCAGTCCTAGGTATAATACTAGTTAACGGTTAAATAGCCCGATGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGT\n",
    "```\n",
    "\n",
    "It's read1 is calculated as: ``TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT``\n",
    "\n",
    "It's read2 as: ``AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA``\n",
    "\n",
    "The second bin, bin 1, labeled 1000, is associated with ``AACCGCAA``. This is from the unique barcode file.\n",
    "\n",
    "Based on the binning file, poxBp_T7_TGT has 32 reads in the second bin. So I generated 32 records for read1 and 32 records for read2. \n",
    "\n",
    "The 32 records in read1 look like this:\n",
    "\n",
    "```\n",
    "@SIM:1:FCX:1:14:6329:1045:1:N:0:AACCGCAA\n",
    "TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT\n",
    "+\n",
    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
    "```\n",
    "\n",
    "and the 32 records in read2 look like this:\n",
    "\n",
    "```\n",
    "@SIM:1:FCX:1:14:6329:1045:2:N:0:AACCGCAA\n",
    "AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA\n",
    "+\n",
    "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n",
    "```\n",
    "\n",
    "Note the read is the sequence, and the bin index goes in the index field of the header. The read field is different between the two files.\n",
    "\n",
    "I did this for each of the guide variants and each of the bins to get a total of 25980000 records per file.\n",
    "\n",
    "I was able to read the files with Biopythons SeqIO in the order of minutes.\n",
    "\n",
    "#### Other Features of this Notebook\n",
    "Provides several progress \"bar\" to show user how long the user has to wait. More of a sanity check to show the program isn't broken and is indeed making progress."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from IPython.display import clear_output\n",
    "from Bio import SeqIO "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parameters\n",
    "In order for this code to run, you must provide a few parameters.\n",
    "* ``seq_name``: the file name to be read. It should have a column ID which is unique for every guide variant.\n",
    "* ``barcode_fname``: the txt file that contains all the barcodes. Each barcode is on it's own line.\n",
    "* ``bin_fname``: the file name that has the generated mock bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_fname = 'data/20191118-Sequences for synthesis_v2_VALIDATED.xlsx'\n",
    "barcode_fname = 'data/8nt_index_dist4_barcode.txt'\n",
    "bin_fname = 'data/mock_bins_counts.xlsx'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading in Files\n",
    "In this section, we read in the files and also show an example of the data we get"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library</th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Spacer</th>\n",
       "      <th>PAM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T7_TGT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>TCCGTCAGATGAACTAAACT</td>\n",
       "      <td>TGT</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T17_GAA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CCCTTCCCCCTCCGTCAGAT</td>\n",
       "      <td>GAA</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T18_TGA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>TCCCTTCCCCCTCCGTCAGA</td>\n",
       "      <td>TGA</td>\n",
       "      <td>18.0</td>\n",
       "      <td>Top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T20_GAT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CATCCCTTCCCCCTCCGTCA</td>\n",
       "      <td>GAT</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T21_AGA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CCATCCCTTCCCCCTCCGTC</td>\n",
       "      <td>AGA</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Top</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Library             ID                                           Sequence  \\\n",
       "0  1_poxB   poxBp_T7_TGT  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "1  1_poxB  poxBp_T17_GAA  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "2  1_poxB  poxBp_T18_TGA  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "3  1_poxB  poxBp_T20_GAT  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "4  1_poxB  poxBp_T21_AGA  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "\n",
       "                 Spacer  PAM  Position Strand  \n",
       "0  TCCGTCAGATGAACTAAACT  TGT       7.0    Top  \n",
       "1  CCCTTCCCCCTCCGTCAGAT  GAA      17.0    Top  \n",
       "2  TCCCTTCCCCCTCCGTCAGA  TGA      18.0    Top  \n",
       "3  CATCCCTTCCCCCTCCGTCA  GAT      20.0    Top  \n",
       "4  CCATCCCTTCCCCCTCCGTC  AGA      21.0    Top  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_seq\n",
    "# a pandas df where we load the file of sequences\n",
    "df_seq = pd.read_excel(seq_fname)\n",
    "df_seq = df_seq.drop(columns=['Unnamed: 0'])\n",
    "\n",
    "df_seq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AACCAACG',\n",
       " 'AACCGCAA',\n",
       " 'AACCTGGC',\n",
       " 'AACGACGT',\n",
       " 'AACGCATC',\n",
       " 'AACGTTAG',\n",
       " 'AACTAGTA',\n",
       " 'AACTCTCT',\n",
       " 'AAGAACCA',\n",
       " 'AAGACGAC',\n",
       " 'AAGAGAGT',\n",
       " 'AAGCATAT',\n",
       " 'AAGCTATA',\n",
       " 'AAGGCTGA',\n",
       " 'AAGTCCTG',\n",
       " 'AATAATGC',\n",
       " 'AATAGGCG',\n",
       " 'AATATCAT',\n",
       " 'AATCCGTT',\n",
       " 'AATTGAAC',\n",
       " 'ACCAAGAG',\n",
       " 'ACCAGACC',\n",
       " 'ACCGTCCA',\n",
       " 'ACCTCAGG',\n",
       " 'ACCTGCTT',\n",
       " 'ACGACTCG',\n",
       " 'ACGATCGC',\n",
       " 'ACGCAGTC',\n",
       " 'ACGCGAAG',\n",
       " 'ACGGAACT',\n",
       " 'ACGTTGAT',\n",
       " 'ACTATATG',\n",
       " 'ACTCCTAC',\n",
       " 'ACTGGAGA',\n",
       " 'ACTTACCG',\n",
       " 'AGAACGCT',\n",
       " 'AGAAGCGG',\n",
       " 'AGAATAAC',\n",
       " 'AGACCATG',\n",
       " 'AGAGATCG',\n",
       " 'AGAGTCTT',\n",
       " 'AGATAAGA',\n",
       " 'AGATGGTC',\n",
       " 'AGCAACTC',\n",
       " 'AGCAGTAT',\n",
       " 'AGCTTGCG',\n",
       " 'AGGCGCCT',\n",
       " 'AGGCTTGG',\n",
       " 'AGTATGGA',\n",
       " 'AGTTCCGC',\n",
       " 'ATAACTTC',\n",
       " 'ATACGAGC',\n",
       " 'ATACTCAG',\n",
       " 'ATAGAGAC',\n",
       " 'ATATTACT',\n",
       " 'ATGATTAA',\n",
       " 'ATGGCCAT',\n",
       " 'ATTAGCTA',\n",
       " 'ATTCATTG',\n",
       " 'ATTGGTCC',\n",
       " 'CAACGGAC',\n",
       " 'CAAGAGTT',\n",
       " 'CAAGCAGG',\n",
       " 'CAAGTCAA',\n",
       " 'CAATGCCT',\n",
       " 'CAATTATC',\n",
       " 'CAGATTCC',\n",
       " 'CAGCAAGC',\n",
       " 'CAGCGTTG',\n",
       " 'CAGTCGGT',\n",
       " 'CATACCTC',\n",
       " 'CATCTCCG',\n",
       " 'CATTCTAA',\n",
       " 'CCAACCAT',\n",
       " 'CCAATTGG',\n",
       " 'CCAGCGCC',\n",
       " 'CCATACGC',\n",
       " 'CCGAGGAA',\n",
       " 'CCGCCATT',\n",
       " 'CCGGACTG',\n",
       " 'CCGGTAAC',\n",
       " 'CCTAATCT',\n",
       " 'CCTCGCGT',\n",
       " 'CCTGGTAG',\n",
       " 'CGAAGATT',\n",
       " 'CGACCTGC',\n",
       " 'CGATAGAG',\n",
       " 'CGCAAGGT',\n",
       " 'CGCATCAG',\n",
       " 'CGCCGAGG',\n",
       " 'CGCCTTCT',\n",
       " 'CGCGATAC',\n",
       " 'CGCTCAAT',\n",
       " 'CGGCTCTC',\n",
       " 'CGTCAGCC',\n",
       " 'CGTTACTT',\n",
       " 'CTAGGTTA',\n",
       " 'CTATCTCG',\n",
       " 'CTCAGGTC',\n",
       " 'CTCGAGCG',\n",
       " 'CTCGCCGC',\n",
       " 'CTCGTATT',\n",
       " 'CTCTTGAA',\n",
       " 'CTGAGACG',\n",
       " 'CTTATAGC',\n",
       " 'CTTCCAAG',\n",
       " 'GAAGTTGC',\n",
       " 'GACCGATT',\n",
       " 'GACTGCGC',\n",
       " 'GATGCCAG',\n",
       " 'GCAGCTTG',\n",
       " 'GCAGGAAT',\n",
       " 'GCATATAA',\n",
       " 'GCATGGCG',\n",
       " 'GCCATAGT',\n",
       " 'GCCTCGAC',\n",
       " 'GCGAGTTC',\n",
       " 'GCTCTACC',\n",
       " 'GCTGCGGT',\n",
       " 'GGACGTAG',\n",
       " 'GGCCAGAA',\n",
       " 'GGCGAATG',\n",
       " 'GGTACTCC',\n",
       " 'GGTCAAGT',\n",
       " 'GGTTCGTA',\n",
       " 'GTAACGAG',\n",
       " 'GTAGACCT',\n",
       " 'GTCAACGG',\n",
       " 'GTCCTTAC',\n",
       " 'GTCTAACC',\n",
       " 'GTTATGCT',\n",
       " 'TAATTCGG',\n",
       " 'TAGTAATT',\n",
       " 'TATCAGAG',\n",
       " 'TATGGACT',\n",
       " 'TCAGTACG',\n",
       " 'TCCGGCGG',\n",
       " 'TCGGATGC',\n",
       " 'TCTATGAC',\n",
       " 'TCTTCATC',\n",
       " 'TGCGCGAG',\n",
       " 'TGCGTAGC',\n",
       " 'TGGAATAG',\n",
       " 'TGGAGGCC',\n",
       " 'TTACCGGT',\n",
       " 'TTATGAAG',\n",
       " 'TTCCTCTA',\n",
       " 'TTCTGGCT',\n",
       " 'TTGAAGGA',\n",
       " 'TTGCGTCA',\n",
       " 'TTGCTAAT',\n",
       " 'TTGGCATG']"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#bar_lst\n",
    "# a list where we keep the barcodes\n",
    "f = open(barcode_fname, \"r\")\n",
    "bar_list = f.read().split('\\n')\n",
    "f.close()\n",
    "\n",
    "bar_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>0</th>\n",
       "      <th>1000</th>\n",
       "      <th>2000</th>\n",
       "      <th>3000</th>\n",
       "      <th>4000</th>\n",
       "      <th>5000</th>\n",
       "      <th>6000</th>\n",
       "      <th>7000</th>\n",
       "      <th>8000</th>\n",
       "      <th>9000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>poxBp_T7_TGT</td>\n",
       "      <td>6</td>\n",
       "      <td>559</td>\n",
       "      <td>5172</td>\n",
       "      <td>7461</td>\n",
       "      <td>1741</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>poxBp_T17_GAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96</td>\n",
       "      <td>1938</td>\n",
       "      <td>7422</td>\n",
       "      <td>4942</td>\n",
       "      <td>586</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>poxBp_T18_TGA</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>1298</td>\n",
       "      <td>5064</td>\n",
       "      <td>6162</td>\n",
       "      <td>2148</td>\n",
       "      <td>230</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>poxBp_T20_GAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>1576</td>\n",
       "      <td>6170</td>\n",
       "      <td>5734</td>\n",
       "      <td>1354</td>\n",
       "      <td>69</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>poxBp_T21_AGA</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>274</td>\n",
       "      <td>2858</td>\n",
       "      <td>6992</td>\n",
       "      <td>4239</td>\n",
       "      <td>606</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              ID  0  1000  2000  3000  4000  5000  6000  7000  8000  9000\n",
       "0   poxBp_T7_TGT  6   559  5172  7461  1741    61     0     0     0     0\n",
       "1  poxBp_T17_GAA  0     0     3    96  1938  7422  4942   586    13     0\n",
       "2  poxBp_T18_TGA  0     1    93  1298  5064  6162  2148   230     4     0\n",
       "3  poxBp_T20_GAT  0     0     1    93  1576  6170  5734  1354    69     3\n",
       "4  poxBp_T21_AGA  0     9   274  2858  6992  4239   606    22     0     0"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_bins\n",
    "# a pandas df with the bins read in\n",
    "df_bin = pd.read_excel(bin_fname)\n",
    "df_bin = df_bin.drop(columns=['Unnamed: 0'])\n",
    "\n",
    "df_bin.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Helper Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rev_comp(seq):\n",
    "    '''\n",
    "    Switches a seqence from top to bottom\n",
    "    (ie 'ATGC' becomes 'TACG')    \n",
    "    ARGS:\n",
    "        -seq: (str) a string which represents the top strand of DNA.\n",
    "               Should only have the characters ATGC\n",
    "    RETURNS:\n",
    "        string sequence that represents the bottom sequence\n",
    "        (all Ts become As, As become Ts, Gs become Cs, Cs become Gs)\n",
    "        and is revered\n",
    "        Resulting string will be all uppercase\n",
    "    '''\n",
    "    res = seq.lower()\n",
    "    res = res.replace('t', 'A')\n",
    "    res = res.replace('a', 'T')\n",
    "    res = res.replace('g', 'C')\n",
    "    res = res.replace('c', 'G')\n",
    "    res = res[::-1]\n",
    "    return res\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepping data structures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# num bins:\n",
    "# the number of bins that we used,\n",
    "# should come from the df_bins file\n",
    "# subtract the one to get rid of the label column\n",
    "n_bins = len(list(df_bin.columns)) - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# barcodes: a list of barcodes that we are going to use\n",
    "# we need as many barcodes as there are bins\n",
    "barcodes = bar_list[:n_bins]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library</th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Spacer</th>\n",
       "      <th>PAM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>0</th>\n",
       "      <th>1000</th>\n",
       "      <th>2000</th>\n",
       "      <th>3000</th>\n",
       "      <th>4000</th>\n",
       "      <th>5000</th>\n",
       "      <th>6000</th>\n",
       "      <th>7000</th>\n",
       "      <th>8000</th>\n",
       "      <th>9000</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T7_TGT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>TCCGTCAGATGAACTAAACT</td>\n",
       "      <td>TGT</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>6</td>\n",
       "      <td>559</td>\n",
       "      <td>5172</td>\n",
       "      <td>7461</td>\n",
       "      <td>1741</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T17_GAA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CCCTTCCCCCTCCGTCAGAT</td>\n",
       "      <td>GAA</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96</td>\n",
       "      <td>1938</td>\n",
       "      <td>7422</td>\n",
       "      <td>4942</td>\n",
       "      <td>586</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T18_TGA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>TCCCTTCCCCCTCCGTCAGA</td>\n",
       "      <td>TGA</td>\n",
       "      <td>18.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>1298</td>\n",
       "      <td>5064</td>\n",
       "      <td>6162</td>\n",
       "      <td>2148</td>\n",
       "      <td>230</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T20_GAT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CATCCCTTCCCCCTCCGTCA</td>\n",
       "      <td>GAT</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>1576</td>\n",
       "      <td>6170</td>\n",
       "      <td>5734</td>\n",
       "      <td>1354</td>\n",
       "      <td>69</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T21_AGA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CCATCCCTTCCCCCTCCGTC</td>\n",
       "      <td>AGA</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>274</td>\n",
       "      <td>2858</td>\n",
       "      <td>6992</td>\n",
       "      <td>4239</td>\n",
       "      <td>606</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1727</th>\n",
       "      <td>25_J3</td>\n",
       "      <td>J3_B195_CGC</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>ATAAGCGCTGCGTGAATGAT</td>\n",
       "      <td>CGC</td>\n",
       "      <td>195.0</td>\n",
       "      <td>Bottom</td>\n",
       "      <td>0</td>\n",
       "      <td>83</td>\n",
       "      <td>1363</td>\n",
       "      <td>5877</td>\n",
       "      <td>6013</td>\n",
       "      <td>1569</td>\n",
       "      <td>94</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1728</th>\n",
       "      <td>25_J3</td>\n",
       "      <td>J3_B197_CAA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>AAGCGCTGCGTGAATGATCG</td>\n",
       "      <td>CAA</td>\n",
       "      <td>197.0</td>\n",
       "      <td>Bottom</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>495</td>\n",
       "      <td>5398</td>\n",
       "      <td>7681</td>\n",
       "      <td>1393</td>\n",
       "      <td>29</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1729</th>\n",
       "      <td>25_J3</td>\n",
       "      <td>J3_B201_TGC</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>GCTGCGTGAATGATCGCAAA</td>\n",
       "      <td>TGC</td>\n",
       "      <td>201.0</td>\n",
       "      <td>Bottom</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>277</td>\n",
       "      <td>2954</td>\n",
       "      <td>7155</td>\n",
       "      <td>4063</td>\n",
       "      <td>526</td>\n",
       "      <td>15</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1730</th>\n",
       "      <td>25_J3</td>\n",
       "      <td>J3_OT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>GACTTGTGAAATGCGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>194</td>\n",
       "      <td>3134</td>\n",
       "      <td>7945</td>\n",
       "      <td>3457</td>\n",
       "      <td>262</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1731</th>\n",
       "      <td>25_J3</td>\n",
       "      <td>J3_OT_mutTATA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>GACTTGTGAAATGCGAGACC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>18</td>\n",
       "      <td>936</td>\n",
       "      <td>6037</td>\n",
       "      <td>6672</td>\n",
       "      <td>1292</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1732 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Library             ID  \\\n",
       "0     1_poxB   poxBp_T7_TGT   \n",
       "1     1_poxB  poxBp_T17_GAA   \n",
       "2     1_poxB  poxBp_T18_TGA   \n",
       "3     1_poxB  poxBp_T20_GAT   \n",
       "4     1_poxB  poxBp_T21_AGA   \n",
       "...      ...            ...   \n",
       "1727   25_J3    J3_B195_CGC   \n",
       "1728   25_J3    J3_B197_CAA   \n",
       "1729   25_J3    J3_B201_TGC   \n",
       "1730   25_J3          J3_OT   \n",
       "1731   25_J3  J3_OT_mutTATA   \n",
       "\n",
       "                                               Sequence                Spacer  \\\n",
       "0     CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  TCCGTCAGATGAACTAAACT   \n",
       "1     CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  CCCTTCCCCCTCCGTCAGAT   \n",
       "2     CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  TCCCTTCCCCCTCCGTCAGA   \n",
       "3     CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  CATCCCTTCCCCCTCCGTCA   \n",
       "4     CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  CCATCCCTTCCCCCTCCGTC   \n",
       "...                                                 ...                   ...   \n",
       "1727  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  ATAAGCGCTGCGTGAATGAT   \n",
       "1728  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  AAGCGCTGCGTGAATGATCG   \n",
       "1729  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  GCTGCGTGAATGATCGCAAA   \n",
       "1730  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  GACTTGTGAAATGCGAGACC   \n",
       "1731  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...  GACTTGTGAAATGCGAGACC   \n",
       "\n",
       "      PAM  Position  Strand  0  1000  2000  3000  4000  5000  6000  7000  \\\n",
       "0     TGT       7.0     Top  6   559  5172  7461  1741    61     0     0   \n",
       "1     GAA      17.0     Top  0     0     3    96  1938  7422  4942   586   \n",
       "2     TGA      18.0     Top  0     1    93  1298  5064  6162  2148   230   \n",
       "3     GAT      20.0     Top  0     0     1    93  1576  6170  5734  1354   \n",
       "4     AGA      21.0     Top  0     9   274  2858  6992  4239   606    22   \n",
       "...   ...       ...     ... ..   ...   ...   ...   ...   ...   ...   ...   \n",
       "1727  CGC     195.0  Bottom  0    83  1363  5877  6013  1569    94     1   \n",
       "1728  CAA     197.0  Bottom  0     3   495  5398  7681  1393    29     1   \n",
       "1729  TGC     201.0  Bottom  0     0    10   277  2954  7155  4063   526   \n",
       "1730  NaN       NaN     NaN  0     0     2   194  3134  7945  3457   262   \n",
       "1731  NaN       NaN     NaN  0     0     0    18   936  6037  6672  1292   \n",
       "\n",
       "      8000  9000  \n",
       "0        0     0  \n",
       "1       13     0  \n",
       "2        4     0  \n",
       "3       69     3  \n",
       "4        0     0  \n",
       "...    ...   ...  \n",
       "1727     0     0  \n",
       "1728     0     0  \n",
       "1729    15     0  \n",
       "1730     6     0  \n",
       "1731    45     0  \n",
       "\n",
       "[1732 rows x 17 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df:\n",
    "# join df_seq and df_bins on 'ID'\n",
    "df = df_seq.merge(df_bin, left_on='ID', right_on='ID')\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read Functions\n",
    "We get the 2 reads from this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCTAGCTCAGTCCTAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGT'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.iloc[0, 2] #== 'CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCTAGCTCAGTCCTAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTTTTAGAGCTAGAAATAGCAAGTTAAAATAAGGCTAGTCCGTTATCAACTTGAAAAAGT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_read1(seq):\n",
    "    '''\n",
    "    ARGS:\n",
    "        seq: A string ATCG sequence we will use\n",
    "    Returns:\n",
    "        a string with read1\n",
    "        which is 39 bases; begins with: TAGG (TATAATACTAGT) N20 GTT \n",
    "    '''\n",
    "    #  \n",
    "    i = seq.find('TAGG')\n",
    "    return seq[i:i+39]\n",
    "    \n",
    "seq = df.iloc[0, 2]\n",
    "get_read1(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def get_read2(seq):\n",
    "    '''\n",
    "    ARGS:\n",
    "        seq: A string ATCG sequence we will use\n",
    "    Returns:\n",
    "        a string with read2\n",
    "        which is 36 bases: reverse complement of N20, reverse complement of (sequence) CCTA\n",
    "    '''\n",
    "    seq_rev = rev_comp(seq)\n",
    "    i = seq_rev.find('CCTA')\n",
    "    return seq_rev[i-32:i+4]\n",
    "\n",
    "get_read2(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(36, 36)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(get_read2(seq)), len('ACGGAGCGTTCTGGACACAAACTAGTATTATACCTA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we apply both read1 and read2 into the dataframe so that read1 and read2 for each sequence is known"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Library</th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Spacer</th>\n",
       "      <th>PAM</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>0</th>\n",
       "      <th>1000</th>\n",
       "      <th>2000</th>\n",
       "      <th>3000</th>\n",
       "      <th>4000</th>\n",
       "      <th>5000</th>\n",
       "      <th>6000</th>\n",
       "      <th>7000</th>\n",
       "      <th>8000</th>\n",
       "      <th>9000</th>\n",
       "      <th>read1</th>\n",
       "      <th>read2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T7_TGT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>TCCGTCAGATGAACTAAACT</td>\n",
       "      <td>TGT</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>6</td>\n",
       "      <td>559</td>\n",
       "      <td>5172</td>\n",
       "      <td>7461</td>\n",
       "      <td>1741</td>\n",
       "      <td>61</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT</td>\n",
       "      <td>AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T17_GAA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CCCTTCCCCCTCCGTCAGAT</td>\n",
       "      <td>GAA</td>\n",
       "      <td>17.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>96</td>\n",
       "      <td>1938</td>\n",
       "      <td>7422</td>\n",
       "      <td>4942</td>\n",
       "      <td>586</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>TAGGTATAATACTAGTCCCTTCCCCCTCCGTCAGATGTT</td>\n",
       "      <td>ATCTGACGGAGGGGGAAGGGACTAGTATTATACCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T18_TGA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>TCCCTTCCCCCTCCGTCAGA</td>\n",
       "      <td>TGA</td>\n",
       "      <td>18.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>1298</td>\n",
       "      <td>5064</td>\n",
       "      <td>6162</td>\n",
       "      <td>2148</td>\n",
       "      <td>230</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>TAGGTATAATACTAGTTCCCTTCCCCCTCCGTCAGAGTT</td>\n",
       "      <td>TCTGACGGAGGGGGAAGGGAACTAGTATTATACCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T20_GAT</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CATCCCTTCCCCCTCCGTCA</td>\n",
       "      <td>GAT</td>\n",
       "      <td>20.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>93</td>\n",
       "      <td>1576</td>\n",
       "      <td>6170</td>\n",
       "      <td>5734</td>\n",
       "      <td>1354</td>\n",
       "      <td>69</td>\n",
       "      <td>3</td>\n",
       "      <td>TAGGTATAATACTAGTCATCCCTTCCCCCTCCGTCAGTT</td>\n",
       "      <td>TGACGGAGGGGGAAGGGATGACTAGTATTATACCTA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1_poxB</td>\n",
       "      <td>poxBp_T21_AGA</td>\n",
       "      <td>CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...</td>\n",
       "      <td>CCATCCCTTCCCCCTCCGTC</td>\n",
       "      <td>AGA</td>\n",
       "      <td>21.0</td>\n",
       "      <td>Top</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>274</td>\n",
       "      <td>2858</td>\n",
       "      <td>6992</td>\n",
       "      <td>4239</td>\n",
       "      <td>606</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>TAGGTATAATACTAGTCCATCCCTTCCCCCTCCGTCGTT</td>\n",
       "      <td>GACGGAGGGGGAAGGGATGGACTAGTATTATACCTA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Library             ID                                           Sequence  \\\n",
       "0  1_poxB   poxBp_T7_TGT  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "1  1_poxB  poxBp_T17_GAA  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "2  1_poxB  poxBp_T18_TGA  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "3  1_poxB  poxBp_T20_GAT  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "4  1_poxB  poxBp_T21_AGA  CTGAAGTCAGCCCCATACGATATAAGTTGTTACTAGATTGACAGCT...   \n",
       "\n",
       "                 Spacer  PAM  Position Strand  0  1000  2000  3000  4000  \\\n",
       "0  TCCGTCAGATGAACTAAACT  TGT       7.0    Top  6   559  5172  7461  1741   \n",
       "1  CCCTTCCCCCTCCGTCAGAT  GAA      17.0    Top  0     0     3    96  1938   \n",
       "2  TCCCTTCCCCCTCCGTCAGA  TGA      18.0    Top  0     1    93  1298  5064   \n",
       "3  CATCCCTTCCCCCTCCGTCA  GAT      20.0    Top  0     0     1    93  1576   \n",
       "4  CCATCCCTTCCCCCTCCGTC  AGA      21.0    Top  0     9   274  2858  6992   \n",
       "\n",
       "   5000  6000  7000  8000  9000                                    read1  \\\n",
       "0    61     0     0     0     0  TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT   \n",
       "1  7422  4942   586    13     0  TAGGTATAATACTAGTCCCTTCCCCCTCCGTCAGATGTT   \n",
       "2  6162  2148   230     4     0  TAGGTATAATACTAGTTCCCTTCCCCCTCCGTCAGAGTT   \n",
       "3  6170  5734  1354    69     3  TAGGTATAATACTAGTCATCCCTTCCCCCTCCGTCAGTT   \n",
       "4  4239   606    22     0     0  TAGGTATAATACTAGTCCATCCCTTCCCCCTCCGTCGTT   \n",
       "\n",
       "                                  read2  \n",
       "0  AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA  \n",
       "1  ATCTGACGGAGGGGGAAGGGACTAGTATTATACCTA  \n",
       "2  TCTGACGGAGGGGGAAGGGAACTAGTATTATACCTA  \n",
       "3  TGACGGAGGGGGAAGGGATGACTAGTATTATACCTA  \n",
       "4  GACGGAGGGGGAAGGGATGGACTAGTATTATACCTA  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['read1'] = df['Sequence'].apply(get_read1)\n",
    "df['read2'] = df['Sequence'].apply(get_read2)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generate new data\n",
    "We now generate a dataframe with: Read 1; Read 2; Bin Index; "
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "pseudocode:\n",
    "\n",
    "df_fastq = empty\n",
    "For each row in df\n",
    "    For each bin:\n",
    "        barcode = barcodes[bin]\n",
    "        num_reads = value in that bin\n",
    "        for num_reads:\n",
    "            add read1, read2, and barcode to df_fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100.0 %\n"
     ]
    }
   ],
   "source": [
    "d_fastq = {'Read 1': [], 'Read 2': [], 'Bin Index': []}\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    # just for seeing progress:\n",
    "    print(round((index + 1)/len(df) * 100, 1), \"%\")\n",
    "    clear_output(wait=True)\n",
    "    \n",
    "    # for each sequence get read1 and read2\n",
    "    read1 = row['read1']\n",
    "    read2 = row['read2']\n",
    "    \n",
    "    # for each bin in the barcodes\n",
    "    for b in range(n_bins):\n",
    "        # get the number of reads for this bin\n",
    "        num_reads = df.iloc[index, 7 + b]        \n",
    "        barcode = barcodes[b]\n",
    "        \n",
    "        for r in range(num_reads):\n",
    "            d_fastq['Read 1'].append(read1)\n",
    "            d_fastq['Read 2'].append(read2)\n",
    "            d_fastq['Bin Index'].append(barcode)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_fastq = pd.DataFrame.from_dict(d_fastq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Read 1</th>\n",
       "      <th>Read 2</th>\n",
       "      <th>Bin Index</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT</td>\n",
       "      <td>AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA</td>\n",
       "      <td>AACCAACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT</td>\n",
       "      <td>AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA</td>\n",
       "      <td>AACCAACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT</td>\n",
       "      <td>AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA</td>\n",
       "      <td>AACCAACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT</td>\n",
       "      <td>AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA</td>\n",
       "      <td>AACCAACG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT</td>\n",
       "      <td>AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA</td>\n",
       "      <td>AACCAACG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    Read 1  \\\n",
       "0  TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT   \n",
       "1  TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT   \n",
       "2  TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT   \n",
       "3  TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT   \n",
       "4  TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT   \n",
       "\n",
       "                                 Read 2 Bin Index  \n",
       "0  AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA  AACCAACG  \n",
       "1  AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA  AACCAACG  \n",
       "2  AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA  AACCAACG  \n",
       "3  AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA  AACCAACG  \n",
       "4  AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA  AACCAACG  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_fastq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "---\n",
    "\n",
    "# To Fastq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We look at some exmples of fastq datafiles to figure out what we would need to export."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exploring Ban Wang's fastq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['NB501203', '285', 'H2TKCBGXC', '1', '11101', '6053', '1067']"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fname = \"data/501_0_S27_R2_001.fastq\"\n",
    "for r in SeqIO.parse(fname, \"fastq\"):\n",
    "    break\n",
    "# print(r, '\\n')\n",
    "# print(r.seq)\n",
    "# print(r.id)\n",
    "# type(r.id)\n",
    "r.id.split(\":\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@NB501203:285:H2TKCBGXC:1:11101:6053:1067 2:N:0:TCGCCTTA+TCATGAGC\n",
      "GAACGATNNN\n",
      "+\n",
      "AAAAAEE###\n",
      "@NB501203:285:H2TKCBGXC:1:11101:18448:1067 2:N:0:TCGCCTTA+TCATGAGC\n",
      "CCGAAAGNNN\n",
      "+\n",
      "AAAAAEE###\n"
     ]
    }
   ],
   "source": [
    "f = open(fname, \"r\")\n",
    "i = 0\n",
    "for line in f:\n",
    "    print(line, end = \"\")\n",
    "    if i == 7:\n",
    "        break\n",
    "    i += 1\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Looking [here](https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/FileFormat_FASTQ-files_swBS.htm) for documentation on how the illumina file returns data"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "@<instrument>:<run number>:<flowcell ID>:<lane>:<tile>:<x-pos>:<y-pos>:<UMI> <read>:<is filtered>:<control number>:<index>"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Each entry in a FASTQ file consists of 4 lines:\n",
    "\t▶ \tSequence identifier\n",
    "\t▶ \tSequence\n",
    "\t▶ \tQuality score identifier line (consisting only of a +)\n",
    "\t▶ \tQuality score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Raw Record:\n",
      "\n",
      "@SIM:1:FCX:1:15:6329:1045:GATTACT+GTCTTAAC 1:N:0:ATCCGA\n",
      "TCGCACTCAACGCCCTGCATATGACAAGACAGAATC\n",
      "+\n",
      "<>;##=><9=AAAAAAAAAA9#:<#<;<<<????#=\n"
     ]
    }
   ],
   "source": [
    "record = \"@SIM:1:FCX:1:15:6329:1045:GATTACT+GTCTTAAC 1:N:0:ATCCGA\\nTCGCACTCAACGCCCTGCATATGACAAGACAGAATC\\n+\\n<>;##=><9=AAAAAAAAAA9#:<#<;<<<????#=\"\n",
    "print(\"Raw Record:\\n\")\n",
    "print(record)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record Lines:\n",
      "\n",
      "Sequence ID \t @SIM:1:FCX:1:15:6329:1045:GATTACT+GTCTTAAC 1:N:0:ATCCGA\n",
      "Sequence \t TCGCACTCAACGCCCTGCATATGACAAGACAGAATC\n",
      "Quality ID \t +\n",
      "Quality score \t <>;##=><9=AAAAAAAAAA9#:<#<;<<<????#=\n"
     ]
    }
   ],
   "source": [
    "print(\"Record Lines:\\n\")\n",
    "for line, label in zip(record.split(\"\\n\"), ['Sequence ID', 'Sequence', 'Quality ID', 'Quality score']):\n",
    "    print(label, \"\\t\", line)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence ID Elements:\n",
      "\n",
      "instrument \t @SIM\n",
      "run num \t 1\n",
      "flowcell id \t FCX\n",
      "lane\t \t 1\n",
      "title\t \t 15\n",
      "xpos\t \t 6329\n",
      "ypos\t \t 1045\n",
      "UMI/Read \t GATTACT+GTCTTAAC 1\n",
      "is filtered \t N\n",
      "control num \t 0\n",
      "index\t \t ATCCGA\n"
     ]
    }
   ],
   "source": [
    "print(\"Sequence ID Elements:\\n\")\n",
    "seqid = record.split(\"\\n\")[0]\n",
    "labels = ['instrument', 'run num', 'flowcell id', 'lane\\t', 'title\\t', 'xpos\\t', 'ypos\\t', 'UMI/Read', 'is filtered', 'control num', 'index\\t']\n",
    "for e, label in zip(seqid.split(\":\"), labels):\n",
    "    print(label, \"\\t\", e)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper Functions\n",
    "Below we have\n",
    "* gen_record: creates a record for the illumina machine given the fields\n",
    "* parse_id: parses the id of a record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gen_record(instrument = 'SIM',\n",
    "               run_num = '1',\n",
    "               flowcell_id = 'FCX',\n",
    "               lane = '1',\n",
    "               title = '14',\n",
    "               xpos = '6329',\n",
    "               ypos = '1045',\n",
    "               UMI = '',\n",
    "               read = '',\n",
    "               is_filtered = 'N',\n",
    "               control_num = '0',\n",
    "               index = '',\n",
    "               seq = '',\n",
    "               quality_score = None\n",
    "              ):\n",
    "    '''\n",
    "    given the elements of a fastq record will generate it.\n",
    "    ARGS:\n",
    "        - instrument: Instrument ID. Characters allowed: a–z, A–Z, 0–9 and underscore\n",
    "        - run_num: numerical. Run number on instrument\n",
    "        - flow_cell_id: numerical\n",
    "        - lane: numerical. Lane number.\n",
    "        - title: numerical. Title number.\n",
    "        - x_pos: numerical. X coordinate of cluster.\n",
    "        - y_pos: numerical. Y coordinate of cluster.\n",
    "        - UMI: sequence (ATGCN). UMI sequences for read 1 and read2 seperated by +\n",
    "        - read: numerical. Read number. 1 can be single read or Read 2 of paired-end\n",
    "        - is_filtered: Y or N. Y if the read is filtered (did not pass), N otherwise\n",
    "        - control_number: numerical. 0 when none of the control bits are on, otherwise it is an even number.\n",
    "        - index: (ATGCN). index of the read\n",
    "        - seq: (ATGCN). The sequence read\n",
    "        - quality_score: the quality score. If it is left as None, will omit the +\n",
    "    RETURN:\n",
    "        Returns a 2 or 4 lined string that represents how this record would be presented in a fastq file\n",
    "    '''\n",
    "    # make sure everything is a string or in the appropraite form\n",
    "    instrument = str(instrument)\n",
    "    run_num = str(run_num)\n",
    "    flowcell_id = str(flowcell_id)\n",
    "    lane = str(lane)\n",
    "    title = str(title)\n",
    "    xpos = str(xpos)\n",
    "    ypos = str(ypos)\n",
    "    UMI = str(UMI)\n",
    "    read = str(read)\n",
    "    if type(is_filtered) == 'bool':\n",
    "        if is_filtered:\n",
    "            is_filtered = 'Y'\n",
    "        else:\n",
    "            is_filtered = 'N'\n",
    "    is_filtered = str(is_filtered)\n",
    "    control_num = str(control_num)\n",
    "    index = str(index)\n",
    "    seq = str(seq)\n",
    "    \n",
    "    line1 = \"@\" + instrument + \":\" + run_num + ':' + flowcell_id + ':' + lane + ':' + title + ':' + xpos + ':' + ypos + ':' + UMI + read  + ':' + is_filtered + ':' + control_num + ':' + index + \"\\n\"\n",
    "    line2 = seq + \"\\n\"\n",
    "    line3 = \"\" \n",
    "    line4 = \"\"\n",
    "    if not (quality_score is None):\n",
    "        line3 = \"+\\n\"\n",
    "        line4 = str(quality_score) + \"\\n\"\n",
    "    return line1 + line2 + line3 + line4\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'instrument': '@SIM',\n",
       " 'run num': '1',\n",
       " 'flowcell_id': 'FCX',\n",
       " 'lane': '1',\n",
       " 'title': '15',\n",
       " 'xpos': '6329',\n",
       " 'ypos': '1045',\n",
       " 'UMI': 'GATTACT+GTCTTAAC',\n",
       " 'read': '1',\n",
       " 'is_filtered': 'N',\n",
       " 'control_num': '0',\n",
       " 'index': 'ATCCGA'}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def parse_id(record):\n",
    "    '''\n",
    "    Given a whole record (or just the id)\n",
    "    parses it and returns it as a dictionary\n",
    "    ARGS:\n",
    "        - record: (str) a string of the record or just the id\n",
    "    RETURNS:\n",
    "        a dictionary with the following keys:\n",
    "        - instrument: Instrument ID. Characters allowed: a–z, A–Z, 0–9 and underscore\n",
    "        - run_num: numerical. Run number on instrument\n",
    "        - flow_cell_id: numerical\n",
    "        - lane: numerical. Lane number.\n",
    "        - title: numerical. Title number.\n",
    "        - x_pos: numerical. X coordinate of cluster.\n",
    "        - y_pos: numerical. Y coordinate of cluster.\n",
    "        - UMI: sequence (ATGCN). UMI sequences for read 1 and read2 seperated by +\n",
    "        - read: numerical. Read number. 1 can be single read or Read 2 of paired-end\n",
    "        - is_filtered: Y or N. Y if the read is filtered (did not pass), N otherwise\n",
    "        - control_number: numerical. 0 when none of the control bits are on, otherwise it is an even number.\n",
    "        - index: (ATGCN). index of the read\n",
    "    '''\n",
    "    seqid = record.split(\"\\n\")[0]\n",
    "    labels = ['instrument', 'run num', 'flowcell_id', 'lane', 'title', 'xpos', 'ypos', 'UMI/read', 'is_filtered', 'control_num', 'index']\n",
    "    d = dict()\n",
    "    for e, label in zip(seqid.split(\":\"), labels):\n",
    "        if label == 'UMI/read':\n",
    "            umi, read = e.split(' ')\n",
    "            d['UMI'] = umi\n",
    "            d['read'] = read\n",
    "        else:\n",
    "            d[label] = e\n",
    "    return d\n",
    "     \n",
    "parse_id(record)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In our data...\n",
    "* Sequence ID\n",
    "    * instrument: @SIM\n",
    "    * run num: NA\n",
    "    * flowcell id: NA\n",
    "    * lane: NA\n",
    "    * title: NA\n",
    "    * xpos: NA\n",
    "    * ypos: NA\n",
    "    * UMI/Read: ?!?!?!?!?!?!?!?!?!?!?!?!?!!??!!??!?!?!\n",
    "    * is filtered: NA\n",
    "    * control num: NA\n",
    "    * index: ?!?!?!?!?!?!?!?!?!?!?!?!?!!??!!??!?!?!\n",
    "* Sequence\n",
    "    * The sequence it read\n",
    "* Quality score id line (+)\n",
    "* Quality score\n",
    "    * placeholder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99.9 %\n"
     ]
    }
   ],
   "source": [
    "fname1 = \"data/read1_barcode_as_index.fastq\"\n",
    "fname2 = \"data/read2_barcode_as_index.fastq\"\n",
    "\n",
    "f1 = open(fname1, \"w\")\n",
    "f2 = open(fname2, \"w\")\n",
    "\n",
    "d_fastq = {'Read 1': [], 'Read 2': [], 'Bin Index': []}\n",
    "\n",
    "for index, row in df.iterrows():\n",
    "    # just for seeing progress:\n",
    "    print(round(index/len(df) * 100, 1), \"%\")\n",
    "    clear_output(wait=True)\n",
    "    \n",
    "    # for each sequence get read1 and read2\n",
    "    read1 = row['read1']\n",
    "    read2 = row['read2']\n",
    "    \n",
    "    # generate random qs\n",
    "    qs1 = \"A\" * len(read1) \n",
    "    qs2 = \"A\" * len(read2)\n",
    "        \n",
    "    # for each bin in the barcodes\n",
    "    for b in range(n_bins):\n",
    "        # get the number of reads for this bin\n",
    "        num_reads = df.iloc[index, 7 + b]        \n",
    "        barcode = barcodes[b]       \n",
    "        \n",
    "        # generate the records\n",
    "        # using defaults provided in the documentation\n",
    "        r1 = gen_record(read = 1, index = barcode, seq = read1, quality_score = qs1)\n",
    "        r2 = gen_record(read = 2, index = barcode, seq = read2, quality_score = qs2)\n",
    "        \n",
    "        for r in range(num_reads):\n",
    "            # write into the two files\n",
    "            f1.write(r1)\n",
    "            f2.write(r2)\n",
    "            \n",
    "f1.close()\n",
    "f2.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Testing to see if that worked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('@SIM:1:FCX:1:14:6329:1045:1:N:0:AAGACGAC\\nTAGGGCGTGTACTAGTGACTTGTGAAATGCGAGACCGTT\\n+\\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\\n',\n",
       " '@SIM:1:FCX:1:14:6329:1045:2:N:0:AAGACGAC\\nGGTCTCGCATTTCACAAGTCACTAGTACACGCCCTA\\n+\\nAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\\n')"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r1, r2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: SIM:1:FCX:1:14:6329:1045:1:N:0:AACCAACG\n",
      "Name: SIM:1:FCX:1:14:6329:1045:1:N:0:AACCAACG\n",
      "Description: SIM:1:FCX:1:14:6329:1045:1:N:0:AACCAACG\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('TAGGTATAATACTAGTTCCGTCAGATGAACTAAACTGTT')\n"
     ]
    }
   ],
   "source": [
    "i = 0\n",
    "for r in SeqIO.parse(fname1, \"fastq\"):\n",
    "    print(r)\n",
    "    i += 1\n",
    "    if i == 1:\n",
    "        break\n",
    "\n",
    "# number of records: 25980000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SIM:1:FCX:1:14:6329:1045:2:N:0:AACCAACG\n",
      "AGTTTAGTTCATCTGACGGAACTAGTATTATACCTA\n",
      "+\n",
      "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\n"
     ]
    }
   ],
   "source": [
    "f = open(fname2, \"r\")\n",
    "i = 0\n",
    "for line in f:\n",
    "    print(line, end = \"\")\n",
    "    if i == 3:\n",
    "        break\n",
    "    i += 1\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
